tests/test_build_linkcheck.py

```diff
@@ -382,3 +392,198 @@ def test_connect_to_selfsigned_nonexistent_cert_file(app):
+def test_too_many_requests_user_retry_float(app, capsys):
+    app.config.linkcheck_retry_on_rate_limit = {
+        "localhost:7777": lambda response, delay: 0.01,
+    }
+    retry_handler = make_retry_after_handler([(429, None), (200, None)], log_time=True)
+    with http_server(retry_handler), \
+         mock.patch("sphinx.builders.linkcheck.QUEUE_POLL_SECS", 1.0):
+        app.builder.build_all()
+    content = (app.outdir / 'output.json').read_text()
+    assert json.loads(content) == {
+        "filename": "index.rst",
+        "lineno": 1,
+        "status": "working",
+        "code": 0,
+        "uri": "http://localhost:7777/",
+        "info": "",
+    }
+    _stdout, stderr = capsys.readouterr()
+    first_req_log, second_req_log = stderr.splitlines()
+    first_timestamp_str = ts_re.match(first_req_log).group("ts")
+    second_timestamp_str = ts_re.match(second_req_log).group("ts")
+    if sys.version_info >= (3, 7):
+        first_timestamp = datetime.fromisoformat(first_timestamp_str)
+        second_timestamp = datetime.fromisoformat(second_timestamp_str)
+    else:
+        first_timestamp = datetime.strptime(first_timestamp_str, "%Y-%m-%dT%H:%M:%S.%f")
+        second_timestamp = datetime.strptime(second_timestamp_str, "%Y-%m-%dT%H:%M:%S.%f")
+    # Duration for which the thread does its work.
+    margin = 1.0
+    assert 1.0 <= (second_timestamp - first_timestamp).total_seconds() < 1.0 + margin
+
+
+@pytest.mark.sphinx('linkcheck', testroot='linkcheck-localserver', freshenv=True)
+def test_too_many_requests_retry_exponential_backoff(app, capsys):
+    retry_handler = make_retry_after_handler(
+        [(429, None), (429, None), (200, None)],
+        log_time=True,
+    )
+    with http_server(retry_handler),\
+         mock.patch("sphinx.builders.linkcheck.DEFAULT_DELAY", 1), \
+         mock.patch("sphinx.builders.linkcheck.QUEUE_POLL_SECS", 0.01):
+        start = datetime.now()
+        app.builder.build_all()
+    content = (app.outdir / 'output.json').read_text()
+    assert json.loads(content) == {
+        "filename": "index.rst",
+        "lineno": 1,
+        "status": "working",
+        "code": 0,
+        "uri": "http://localhost:7777/",
+        "info": "",
+    }
+    _stdout, stderr = capsys.readouterr()
+    req_logs = stderr.splitlines()
+    timestamps = []
+    assert len(req_logs) == 3
+    for line in req_logs:
+        timestamp_str = ts_re.match(line).group("ts")
+        if sys.version_info >= (3, 7):
+            timestamp = datetime.fromisoformat(timestamp_str)
+        else:
+            timestamp = datetime.strptime(timestamp_str, "%Y-%m-%dT%H:%M:%S.%f")
+        timestamps.append(timestamp)
+    # Duration which allocated to the thread to do its work.
+    margin = 1.0
+    assert 1.0 <= (timestamps[1] - start).total_seconds() < 1.0 + margin
+    assert 2.0 <= (timestamps[2] - timestamps[1]).total_seconds() < 2.0 + margin
```

Comment:
How about patching `time.sleep()`? It does not waste real CI time and help us to measure how long the linkcheck module sleeps. I don't know it is available even under threading. But it's fragile to do sleeping in the real world.

Reply:
I agree that sleeping is both fragile and wasting test time. It would like to avoid it, but do not see a way to do it. Mocking `time.time()` leads to very tight coupling with the implementation.

Patching `time.sleep()` cause consumer threads to busy-loop (instead of sleep) until the `next_check` timestamp (the priority of the link). It wastes CPU cycles for no benefits.
We could peek at the priority queue, see the `next_check`s queued, but that’s sensitive to the timing of the test and the consumer threads still need to wait until the `next_check`s.

Another option is to unit test the `limit_rate` function, verify its return value (the next check) for various inputs.
It does not verify the entire system behavior, thus offers less guarantees. It is also more coupled with the implementation. OTOH, it avoids sleeping, some mocking and parsing of the test webserver logs, and allows more extensive testing of `limit_rate`. I pushed a separate commit with what I have in mind.